client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/SegmentKeyTracker.java

```diff
@@ -40,20 +42,38 @@ public SegmentKeyTracker(DataFormat dataFormat, SegmentConsistentHash segmentCon
       segmentStream.forEach(i -> keysPerSegment.set(i, new HashSet<>()));
    }
 
+   private Set<WrappedByteArray> drainKeys() {
+      Set<WrappedByteArray> trackedKeys = new HashSet<>();
+      for (int i = 0; i < keysPerSegment.length(); i++) {
+         Set<WrappedByteArray> keys = keysPerSegment.get(i);
+         if (keys != null) trackedKeys.addAll(keys);
+         keysPerSegment.set(i, null);
+      }
+
+      return trackedKeys;
+   }
+
    public boolean track(byte[] key, short status, ClassWhiteList whitelist) {
+      if (!trackSegments) {
+         return keyOnlyTracker.track(key, status, whitelist);
```

Comment:
not sure how this class is used but... can we have an NPE here? 
you are setting `trackSegments` before instantiating `keyOnlyTracker`

Reply:
This class is not supposed to be called by more than 1 thread, it's involved for each iteration result read from the server